File: qq-bot/commands/report.py
"""Report review commands: view, accept, and appeal mission reports from Bot."""

from __future__ import annotations

from api.web_client import get_client
from context.message_context import MessageContext


async def handle_report(ctx: MessageContext) -> bool:
    """Handle report review commands. Returns True if handled."""
    content = ctx.content

    if content == "查看报告":
        return await _handle_view_report(ctx)

    if content == "通过报告":
        return await _handle_accept_report(ctx)

    if content.startswith("申诉报告"):
        return await _handle_appeal_report(ctx)

    return False


async def _handle_view_report(ctx: MessageContext) -> bool:
    client = get_client()
    if not client:
        await ctx.reply("角色卡系统集成未配置。")
        return True

    group_id = ctx.room_id if ctx.source_type == "group" else None
    result = await client.get_pending_report(ctx.player_id, group_id)

    if not result:
        await ctx.reply("无法连接角色卡系统。")
        return True

    if not result.get("success"):
        await ctx.reply(f"查询失败: {result.get('error', '未知错误')}")
        return True

    report = result.get("report")
    if not report:
        await ctx.reply(result.get("message", "当前没有待处理的报告。"))
        return True

    lines = [
        f"【任务报告】{report['missionName']}",
        f"评级: {report['rating']}",
    ]

    rewards = report.get("myRewards", {})
    reward_lines = []
    commend = rewards.get("commend", 0)
    reprimand = rewards.get("reprimand", 0)
    if commend:
        reward_lines.append(f"  嘉奖 {'+' if commend > 0 else ''}{commend}")
    if reprimand:
        reward_lines.append(f"  申诫 {'+' if reprimand > 0 else ''}{reprimand}")
    if rewards.get("mvp"):
        reward_lines.append("  获得MVP")
    if rewards.get("probation"):
        reward_lines.append("  进入查看期")

    if reward_lines:
        lines.append("您的奖惩:")
        lines.extend(reward_lines)
    else:
        lines.append("您的奖惩: 无")

    status_map = {
        "pending": "待确认",
        "accepted": "已接受",
        "appealing": "申诉中",
        "not_participant": "非参与者",
    }
    my_status = report.get("myStatus", "pending")
    lines.append(f"状态: {status_map.get(my_status, my_status)}")

    all_resp = report.get("allResponses", {})
    lines.append(f"进度: {all_resp.get('accepted', 0)}/{all_resp.get('total', 0)} 已确认")
    if all_resp.get("appealing", 0) > 0:
        lines.append(f"  申诉中: {all_resp['appealing']}人")
    if all_resp.get("pending", 0) > 0:
        lines.append(f"  待确认: {all_resp['pending']}人")

    if my_status == "pending":
        lines.append("\n使用「通过报告」接受或「申诉报告 <原因>」提出异议。")

    await ctx.reply("\n".join(lines))
    return True


async def _handle_accept_report(ctx: MessageContext) -> bool:
    client = get_client()
    if not client:
        await ctx.reply("角色卡系统集成未配置。")
        return True

    group_id = ctx.room_id if ctx.source_type == "group" else None
    result = await client.agent_response(ctx.player_id, group_id, "accept")

    if not result:
        await ctx.reply("无法连接角色卡系统。")
        return True

    if not result.get("success"):
        await ctx.reply(f"操作失败: {result.get('error', '未知错误')}")
        return True

    if result.get("autoFinalized"):
        await ctx.reply("已接受评级。所有特工已确认，奖惩已自动结算!")
    else:
        await ctx.reply("已接受评级，等待其他特工响应。")

    return True


async def _handle_appeal_report(ctx: MessageContext) -> bool:
    client = get_client()
    if not client:
        await ctx.reply("角色卡系统集成未配置。")
        return True

    reason = ctx.content[len("申诉报告"):].strip()
    if not reason:
        await ctx.reply("请提供申诉理由，例如: 申诉报告 评级不合理，我的表现应该更高")
        return True

    group_id = ctx.room_id if ctx.source_type == "group" else None
    result = await client.agent_response(ctx.player_id, group_id, "appeal", reason)

    if not result:
        await ctx.reply("无法连接角色卡系统。")
        return True

    if not result.get("success"):
        await ctx.reply(f"操作失败: {result.get('error', '未知错误')}")
        return True

    await ctx.reply("申诉已提交，请等待经理处理。")
    return True

File: src/commands/report.ts
import type { Context, Session } from 'koishi'
import { rawRoomIdOf } from '../room'
import type { WebClient } from '../service/web-client'
import { NETWORK_ERROR_BLOCK } from '../util/messages'
import { requireBound } from '../util/preconditions'
import { sendQQMarkdown, type QQButton } from '../util/qq-markdown'
import { makeWrap } from '../util/safe-action'

export interface ReportDeps {
  web: WebClient | null
  useMarkdown: boolean
}

export function registerReportCommands(ctx: Context, deps: ReportDeps): void {
  const wrap = makeWrap(ctx, deps.useMarkdown)

  ctx.command('查看报告', '查看当前任务的待处理报告').action(
    wrap(async ({ session }) => {
      if (!session) return
      if (!(await requireBound(session, deps))) return
      const qqOpenid = session.userId!
      const groupId = session.isDirect
        ? undefined
        : rawRoomIdOf(session) ?? undefined

      const r = await deps.web!.getPendingReport(qqOpenid, groupId)
      if (!r) return reply(session, deps, NETWORK_ERROR_BLOCK)
      if (!r.success) return reply(session, deps, `> ${r.error ?? '查询失败'}`)
      if (!r.report) {
        return reply(session, deps, `> ${r.message ?? '当前没有待处理的报告。'}`)
      }

      const rep = r.report
      const lines: string[] = []
      lines.push(`# 待处理报告`)
      lines.push('')
      lines.push(`任务　**${rep.missionName}**`)
      lines.push(`评级　**${rep.rating}**`)
      lines.push(`我的状态　**${myStatusLabel(rep.myStatus)}**`)
      lines.push('')
      const rw = rep.myRewards
      const reward: string[] = []
      if (rw.commend) reward.push(`嘉奖 ${rw.commend > 0 ? '+' : ''}${rw.commend}`)
      if (rw.reprimand)
        reward.push(`申诫 ${rw.reprimand > 0 ? '+' : ''}${rw.reprimand}`)
      if (rw.mvp) reward.push('MVP')
      if (rw.probation) reward.push('察看期')
      lines.push(`待结算　${reward.length ? reward.join(' · ') : '无'}`)
      lines.push('')
      lines.push(
        `进度　${rep.allResponses.accepted}/${rep.allResponses.total} 已确认` +
          (rep.allResponses.appealing
            ? `（申诉中 ${rep.allResponses.appealing}）`
            : ''),
      )

      const buttons: QQButton[][] = []
      if (rep.myStatus === 'pending') {
        buttons.push([
          { label: '接受评级', data: '/通过报告', primary: true, type: 'input', enter: true },
          { label: '填写申诉理由', data: '申诉报告 ', type: 'input' },
        ])
      }
      buttons.push([
        { label: '角色状态', data: '/查询状态', type: 'input', enter: true },
        { label: '任务详情', data: '/查看任务', type: 'input', enter: true },
        { label: '操作菜单', data: '/菜单', type: 'input', enter: true },
      ])
      await reply(session, deps, lines.join('\n'), buttons)
    }),
  )

  ctx.command('通过报告', '接受当前评级').action(
    wrap(async ({ session }) => {
      if (!session) return
      if (!(await requireBound(session, deps))) return
      const qqOpenid = session.userId!
      const groupId = session.isDirect
        ? undefined
        : rawRoomIdOf(session) ?? undefined

      const r = await deps.web!.agentResponse(qqOpenid, groupId ?? null, 'accept')
      if (!r) return reply(session, deps, NETWORK_ERROR_BLOCK)
      if (!r.success) return reply(session, deps, `> ${r.error ?? '操作失败'}`)
      const note = r.autoFinalized ? '（任务报告已完结）' : ''
      await reply(
        session,
        deps,
        `> ${r.message ?? '已接受评级'}${note}`,
        [[
          { label: '角色状态', data: '/查询状态', primary: true, type: 'input', enter: true },
          { label: '任务详情', data: '/查看任务', type: 'input', enter: true },
        ]],
      )
    }),
  )

  ctx.command('申诉报告 <reason:text>', '提出申诉').action(
    wrap(async ({ session }, reason) => {
      if (!session) return
      if (!(await requireBound(session, deps))) return
      const qqOpenid = session.userId!
      if (!reason || !reason.trim()) {
        return reply(session, deps, '**用法**　申诉报告 原因（至少 1 字）')
      }
      const groupId = session.isDirect
        ? undefined
        : rawRoomIdOf(session) ?? undefined

      const r = await deps.web!.agentResponse(
        qqOpenid,
        groupId ?? null,
        'appeal',
        reason.trim(),
      )
      if (!r) return reply(session, deps, NETWORK_ERROR_BLOCK)
      if (!r.success) return reply(session, deps, `> ${r.error ?? '申诉失败'}`)
      await reply(session, deps, `> ${r.message ?? '申诉已提交，等待经理处理。'}`, [[
        { label: '查看报告', data: '/查看报告', primary: true, type: 'input', enter: true },
        { label: '任务详情', data: '/查看任务', type: 'input', enter: true },
      ]])
    }),
  )
}

function myStatusLabel(s: string): string {
  return (
    {
      pending: '待响应',
      accepted: '已接受',
      appealing: '申诉中',
      not_participant: '非参与者',
    } as Record<string, string>
  )[s] ?? s
}

async function reply(
  session: Session,
  deps: ReportDeps,
  content: string,
  buttons?: QQButton[][],
): Promise<void> {
  await sendQQMarkdown(session, content, {
    enabled: deps.useMarkdown,
    buttons,
  })
}
